### ISSUE TITLE
Video FIle Size limitations

### ISSUE THREAD

**Issue by oliverellmers** [Samsung/GearVRf#555]
Hi there,

I am having some issues with how your example implements the loading of of video assets into android media player.

I am trying to load a rather large vertical stereo video from the assets directory the same way in which is demonstrated in the minimal 360 video example, but keep hitting and IOException error. I am guessing it is due to the (not sure if this is till relevant after android 2.3???) file size limits for the assets folder or not?

Is there a workaround for this if the app needs to be distributed with the video embeded? It cannot be loaded into an external directory by the user

**Comment by oliverellmers** [comment]
Okay so I managed to get the video loading buy re-compressing it down to a smaller size -> which is not ideal as I would like to be able to playback the bst quality stereoscopic content I can.

I then came across another issue where playback speeds were incredibly slow! Bare in mind, I am also loading in another  (much smaller) video file from assets for my UI menu. I am not sure if this is what was generating drop in frame rate.

I managed to get back to an acceptable frame rate by modifying the way in which the video assets were loaded into the mediaPlayer object, and am loading them in now from res/raw:

`mMediaPlayer = MediaPlayer.create(mGVRContext.getContext(), R.raw.showreel_1);

        mMediaPlayer.setLooping( false );
        
        video = new GVRVideoSceneObject(mGVRContext, mesh, mMediaPlayer, GVRVideoType.VERTICAL_STEREO);`

This now loads the compressed video into mediaPlayer fine, however when I tried this with the uncompressed version (larger file size), the application would crash on open.

The large file in question plays absolutely perfectly when being loaded from external storage.

Any wise words from anybody would be much appreciated

**Comment by liaxim** [comment]
Hi, can you share the file size that doesn't work and the one that does? What is the resolution of the video?

**Comment by oliverellmers** [comment]
Sure thing, sorry for leaving that out...

The uncompressed file:
2000x2000 pixels
60fps
progressive field order
audio: 48000Hz, Stereo
filesize: 1.32 GB


The compressed file:
2000x2000 pixels
60fps
progressive field order
VBR, 1 pass 10 -12Mbps
AAC, 320 Kbps, 48khtz Stereo
filesize: 196 MB

**Comment by liaxim** [comment]
Is rescaling the video an option? I would say more than 1000x1000 is an overkill considering the pixel area the video is shown onto.

It is very likely something is running out of memory with the 1GB+ file.

**Comment by oliverellmers** [comment]
Hmmm I dont really want to rescale the content as I know what 2k content looks like, and it looks great! cutting it in half would be a dramatic loss in quality...

Are there any recommended compression settings for 2k content?

Cheers

**Comment by liaxim** [comment]
Just asking, you could try it at least and see how it looks in stereo. We will look into the issue but since this requires embedding a 1GB video in the apk it could take some time. What is the exception you get? Can you share the logcat output? Thanks.

**Comment by thomasflynn** [comment]
This problem occur if you try to do this without gvrf?  In other words, as a regular android app using the media player?  Iirc, the hw should be able to playback 2kx2k video just fine.  Maybe it is just a memory issue.  I'm not as familiar with the media player, but is there a way to stream rather than load the whole file in?

**Comment by oliverellmers** [comment]
It seems to work fine when building for gear out of unity - haven't had any issue with the uncompressed file regarding playback or memory there... I just prefer developing with gearvrf as it is a lot less restrictive than unity.

Unfortunately my hands are also tied when it comes to down-scaling - my boss eventually wants to be at 4000x4000 pixels and dropping resolution would be a step backward.

GearVRf has definitely been able to play back 2kx2k at 1+ GB from external storage perfectly fine, but now in the process of wanting to put a free download up on the oculus store, so the content needs to be packaged with the application.

I just tried loading the 1GB+ file from the res/raw folder and the application crashes on start.

LogCat file attached from initial error when attempting to load 1+GB file from assets folder using the method demonstrated in the examples.
[logcat1.txt](https://github.com/Samsung/GearVRf/files/259728/logcat1.txt)
I also notice after recreating that issue, that tap events were no longer being listened to

Looking into alternative compression techniques to see if I can drop the file size without losing too much quality - the 196MB file is ghastly!

Alternatively, a way of not having the file embedded in the assets or raw directory would be to have to embedded in an android extension file (ogg) -> have you guys had any experience with extension files and GearVRf?

On another note, with the embedded file,
 even at the smaller size, I am seeing a lot of dropped frames and video stutter.

**Comment by thomasflynn** [comment]
I'm not an expert at the media player stuff, but I'm taking a look anyway.  I do see there are a bunch of:
05-11 17:22:58.431 28354 28354 E MediaPlayer: Error (-38,0)

messages.  Googling reveals:  http://stackoverflow.com/questions/9008770/media-player-called-in-state-0-error-38-0

That may not be the core of your problem, but it might not be helping either.  I'll keep looking...

**Comment by thomasflynn** [comment]
what file extension is the video?

**Comment by thomasflynn** [comment]
ok, so, i created my own 1.2GB video in mp4 by recording a video on my phone.  added it to the assets directory, pointed the media player to it and also got a crash.  here are the relevant logs:

05-11 12:23:51.039 16607 16658 V MediaPlayer: setListener
05-11 12:23:51.039 16607 16658 E filemap : mmap(0,1298322698) failed: Out of memory
05-11 12:23:51.039 16607 16658 W asset   : create map from entry failed
05-11 12:23:51.049 16607 16658 W System.err: java.io.FileNotFoundException: 20160511_104221.mp4
05-11 12:23:51.049  1314  2709 V BroadcastQueue: [background] Process cur broadcast BroadcastRecord{e279268 u0 android.intent.action.PACKAGE_REMOVED qIdx=4}, state= (APP_RECEIVE) DELIVERED for app ProcessRecord{7cf05ee 16641:com.android.settings/1000}
05-11 12:23:51.059 16607 16658 W System.err: 	at android.content.res.AssetManager.openAssetFd(Native Method)
05-11 12:23:51.059 16607 16658 W System.err: 	at android.content.res.AssetManager.openFd(AssetManager.java:379)
05-11 12:23:51.059 16607 16658 W System.err: 	at org.gearvrf.gvr360video.Minimal360Video.onInit(Minimal360Video.java:56)

key part is that mmap failed because it was out of memory.  there's gotta be a way to stream from the assets directory rather than load the whole file into memory.

**Comment by thomasflynn** [comment]
so, i've been trying this all day.  (with a 1+GB file, the adb install time is around 10 minutes...slowing down iteration time).  There's no way you can make that video just a little shorter?  I seem to not get the out of memory problem if i make the file 500MB.  I do get it at 999MB.  Haven't finished binary searching what size will be accepted.  

I tried passing the file as a Uri as illustrated in these stackoverflow links:
http://stackoverflow.com/questions/8269353/how-to-load-videos-from-assets-folder-to-play-them-with-videoview
http://stackoverflow.com/questions/3028717/how-to-play-videos-in-android-from-assets-folder-or-raw-folder

of course another option is to have your application copy the file to sdcard (since you said that works) on startup and then play it.

**Comment by oliverellmers** [comment]
Hey @thomasflynn  sorry for my slow reply - Im guessing you are on the other side of the world from me (Im in the UK)

I had another look into that log file where you found the error(-38, 0), and that was a schoolboy error of myself not naming a file reference correctly and has been fixed.

I did however manage to get better playback by removing one of the the mediaplayer objects - one I have in my 'home' menu system, and instead using a stereoscopic static image - which again, I dont like as I would like to be abel to have animated content in the menu. The video that was used for that object was small, but perhaps having multiple mediaPlayers in one scene is draining a heap memory - I am going to look at dynamic content swapping and have a single media player for the entire application.

Back to the file and size - it is an .mp4, and currently gettign sufficient playback out of the assets folder with a 600MB file.

I did look into streaming via a URI, but couldnt quite get it working - are you able to share your code on your approach there?

I also have looked into using the videoView but im sure I read somewhere that videeo view is simply a wrapper for mediaplayer with embedded controls, so shouldnt have any impact on performance - if that is what you are suggesting?

i will look further into streaming from assets, and/or copying to ssd. The one hang up i have with copying the file to ssd / external storage is the potential of exposing the content to the user. we work in the medical/pharmaceutical animation realm, where content is very sensitive and we cannot allow for it to be accessed by users (well without at least a bit of hard work and know how). Ill have some more updates for you today.

O.

**Comment by oliverellmers** [comment]
After spending most of today trying to find a solution where I can stream directly from the assets or raw folder, I am at a loss..

I have decided to go the route of saving the file to storage in onCreate() when the application is first opened.

The next time it is opened it checks to see if the file exists, if so, it doesn't need to overwrite.

At this stage I am currently saving this to external storage, just to get the ball rolling.

This is problematic for a store app where content is sensitive as

A. Even if the file location the video is being copied to is is "hidden" - by placing a '.' at the beginning of the file you create - this is easily discoverable by enabling display of hidden files.

B. Updates to the application and content will not remove any data stored in external storage, and will not overwrite files with the same name. Therefore content naming will have to change for each update, only bulking out external storage - some users may not be aware of what it is clogging up all of their storage.

Another problem I have come across by saving the content to storage in onCreate is **again** hitting an out of memory error with the 1.24GB file during copy, crashing the application.
With a smaller file being copies - ~400MB - there is also a noticeable hang the first time the application is opened while copying the file over. In the oculus store submission requirements, there must be no hangs, freezing, crashing etc for it to be accepted. 

**Potential solutions / next areas of exploration**

Save the content to internal internal storage where the content will be removed when the application is removed/uninstalled, and the user does not have access to.

Implement a system (god knows how at this stage) that will overwrite any video content if an update is installed. Any ideas are welcome here.

Move the file copying into an asynchronous task and display some form of loading graphic before the content can be played. This is to avoid any hanging/freezing on create.

**But the problem with memory still persists!**

Regardless of all of this additional carry-on, we are still met with the exact same problem we had in the beginning! We are **still** unable to embed and and read/play large stereoscopic video content into a GearVRf application for Oculus Store submission!!!

**Comment by thomasflynn** [comment]
Yeah, we're in California.  I'll keep looking for solutions today.  FWIW, this doesn't appear to be specific to gvrf, but to using the Android MediaPlayer.  In any case, off-hand, i don't know what the Oculus store policy is about large apks, but I know Google Play prefers to have any large files in a separate binary blob:  http://developer.android.com/google/play/expansion-files.html

again, i'll keep looking.  there's gotta be a way to trick the media player to stream for resource/assets files like it does for /sdcard and urls.  instead, it seems to want to mmap the whole file :(

**Comment by thomasflynn** [comment]
as far as protecting the content, you can maybe look at the DRM support in Android: http://developer.android.com/reference/android/drm/package-summary.html

**Comment by thomasflynn** [comment]
i took an attempt at using MediaDataSource: http://developer.android.com/reference/android/media/MediaDataSource.html

found an example here:
https://medium.com/@jacks205/implementing-your-own-android-mediadatasource-e67adb070731#.350hfijp4

i figured with that, i could manually buffer it.  I was trying to get it to work with the small video first and i couldn't get it to work.  I'll show you what i got, but i got my git tree a little messed up in the process, but you can just focus on the changes in gvr-360video.  here's a link to the tree:  https://github.com/thomasflynn/GearVRf-Demos/tree/bigvideo

it seems like it starts reading from it, but then gives a totally bogus position to start at (3267348384).

**Comment by thomasflynn** [comment]
Ok, figured out why that didn't work.  MediaDataSource needs to be seekable.  I can't just read in the stream sequentially. :(

**Comment by danke-sra** [comment]
@thomasflynn The seek at 3267348384 is quite strange since MediaDataSource needs to specify the file size by implementing getSize(). Mediaplayer should not seek beyond that. If the error persists, maybe we can try another approach.

I think in our remote scripting example, we could play a video from a URL. Since it's an acceptable approach for @oliverellmers, why won't we point to that approach:

To create a media player that plays a URI:

```
        MediaPlayer mediaPlayer = new MediaPlayer();
        try {
            mediaPlayer.setDataSource(filenameOrURL);
            mediaPlayer.prepare();
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
```

After it is successfully prepared, we can call mediaPlayer.start() to play the video. To play it automatically, we can add a MediaPlayer.OnPreparedListener to the media player, and call mediaPlayer.start() from the onPrepared method.

**Comment by gearvrf** [comment]
Yeah, look back a couple days ago, that's exactly what i tried.  It still tried to mmap the file.

**Comment by oliverellmers** [comment]
Hey guys,

Apologies for not getting back on this sooner. Have been tasked with a different problem at work.

Anyway, I have still not managed to get this going, and have set up a work around for the mean time by copying the video to internal storage asynchronously when the application is first loaded, displaying a "preparing content" message while we wait. I have compressed the file down to 700MB which give me about a 5-6 second preparation time.

As the copying is being done asynchronously, there is no halt/lag in my 'home' menu, allowing free movement with the sensors, which will hopefully allow a pass on the oculus store.

I unfortunately still run into an out of memory error if I try to copy a file to storage that is any larger than the 700BM file, suggesting a limitation on the device (I am working with S6).

At this stage, it looks like our troubles are limitations of the devices and android itself, and not directly related to GearVRf.

**Comment by danke-sra** [comment]
Thanks for the update. Glad to hear that you are able to somehow reduce the file size and do the decompression (from APK to app data directory) before playing the video. Please let us know if you see other issues.

**Comment by liaxim** [comment]
@oliverellmers Can we close this issue or you need further help?

**Comment by liaxim** [comment]
Planning to close tomorrow

**Comment by Cyganieszka** [comment]
And update in this topic(trying to display big video file from assets)?

**Comment by thomasflynn** [comment]
I think we pretty much ran into Android limitations with this :(

**Comment by Cyganieszka** [comment]
It's a little bit off-topic, but can be very helpful for people seeking answer about video size limitations.

Why, in the whole gearvr documentation there is no information about maximum APK size ? Why you can upload much bigger APK to oculus store, and without any problems test it on alpha/beta channels?

Like a lot of people in this topic, I tried to load big video file from assets. I faced a lot of problems, but in the end I figured the way it works. Application passed content review(after a lot of other problems). And now, I received information from support, that APK size limit is 500MB (even if oculus store accepts bigger files, and it passes content review...). 

**tl;dr** : Don't try to solve problems with playing big files from assets, APK size is limited to 500MB.
**hint** : Gear VR apps have OBB support! (but only uploading by CLI, not a web panel)
